### PR TITLE
I2C: Enable intel's arch/drivers/boards to DTS

### PR DESCRIPTION
**Pull request by tbursztyka**
galileo, all the quark_se(_ss) devboard and the arduino_101(_sss)

### CODE REVIEW

File: dts/bindings/i2c/synopsys,dw-i2c.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: DesignWare SPI node description
+id: synopsys,dw-i2c
```
Comment (galak) [inline dts/bindings/i2c/synopsys,dw-i2c.yaml]: We should use/conform to the linux binding Documentation/devicetree/bindings/i2c/i2c-designware.txt (for things like compat, etc)

File: dts/x86/quark_x1000.dtsi [hunk 2]
```diff
@@ -59,5 +59,16 @@
 
 			status = "disabled";
 		};
+
+		i2c0: i2c@0x90007000 {
+			compatible = "synopsys,dw-i2c";
```
Comment (galak) [inline dts/x86/quark_x1000.dtsi]: Should have #address-cells & #size-cells since its a bus

**Review by galak — CHANGES_REQUESTED** [review]
* Need addr/size cell in all dts
* we need to fix clock frequency / default_cfg:

Take a look at commit 54933b38339dcfa5efcfc3b47d35b197699b2eb9 for example

File: dts/x86/intel_curie.dtsi [hunk 3]
```diff
@@ -92,5 +92,23 @@
...
 			label = "CDC_ACM";
 		};
+
+		i2c0: i2c@0xb0002800 {
+			compatible = "intel,qmsi-i2c";
+			reg = <0xb0002800 0x400>;
+			clock-frequency = <0x12>;
+			label = "I2C_0";
+
+			status = "disabled";
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: Should have #address-cells & #size-cells since its a bus

File: dts/x86/intel_curie.dtsi [hunk 4]
```diff
@@ -92,5 +92,23 @@
...
+		};
+
+		i2c1: i2c@0xb0002c00 {
+			compatible = "intel,qmsi-i2c";
+			reg = <0xb0002c00 0x400>;
+			clock-frequency = <0x12>;
+			label = "I2C_1";
+
```
Comment (galak) [inline dts/x86/intel_curie.dtsi]: Should have #address-cells & #size-cells since its a bus

File: dts/x86/intel_quark_d2000.dtsi [hunk 5]
```diff
@@ -77,5 +77,14 @@
 			gpio-controller;
 			#gpio-cells = <2>;
 		};
+
+		i2c0: i2c@0xb0002800 {
+			compatible = "intel,qmsi-i2c";
+			reg = <0xb0002800 0x400>;
+			clock-frequency = <0x12>;
+			label = "I2C_0";
+
```
Comment (galak) [inline dts/x86/intel_quark_d2000.dtsi]: Should have #address-cells & #size-cells since its a bus

File: arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series [hunk 6]
```diff
@@ -24,7 +24,10 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series]: We should move this in arch/arm/soc/st_stm32/common/Kconfig.defconfig.series
```
if I2C && (I2C_1 || I2C_2 || I2C_3)
config I2C_STM32
	def_bool y
endif # I2C
```
and leave here
```
if I2C_STM32
config I2C_STM32_V2
 	def_bool y
endif # I2C_STM32
```

File: arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series [hunk 7]
```diff
@@ -21,7 +21,10 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V1
+config I2C_STM32
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series]: Same here

File: arch/arm/soc/st_stm32/stm32f3/Kconfig.defconfig.series [hunk 8]
```diff
@@ -21,7 +21,10 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f3/Kconfig.defconfig.series]: Same here

File: arch/arm/soc/st_stm32/stm32f4/Kconfig.defconfig.series [hunk 9]
```diff
@@ -47,7 +47,10 @@ endif #USB
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V1
+config I2C_STM32
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f4/Kconfig.defconfig.series]: Same here

File: arch/arm/soc/st_stm32/stm32l4/Kconfig.defconfig.series [hunk 10]
```diff
@@ -15,7 +15,10 @@ config SOC_SERIES
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
 
-config  I2C_STM32_V2
+config I2C_STM32
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32l4/Kconfig.defconfig.series]: Same here

File: dts/x86/quark_x1000.dtsi [hunk 11]
```diff
@@ -59,5 +59,17 @@
 
 			status = "disabled";
 		};
+
+		i2c0: i2c@0x90007000 {
+			compatible = "snps,designware-i2c";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x90007000 0x400>;
+			interrupts = <18 IRQ_TYPE_LEVEL_LOW>;
```
Comment (ydamigos) [inline dts/x86/quark_x1000.dtsi]: Is IRQ_TYPE_LEVEL_LOW the IRQ priority?

File: arch/x86/soc/intel_quark/quark_x1000/Kconfig.defconfig.series [hunk 12]
```diff
@@ -70,12 +70,11 @@ config I2C_0
 
 if I2C_0
 
-config I2C_0_NAME
-	default "I2C_0"
 config I2C_0_DEFAULT_CFG
 	default 0x12
 config I2C_0_IRQ_PRI
```
Comment (ydamigos) [inline arch/x86/soc/intel_quark/quark_x1000/Kconfig.defconfig.series]: You can get IRQ_PRI from DT.

File: arch/x86/soc/intel_quark/quark_se/Kconfig.defconfig.series [hunk 13]
```diff
@@ -86,6 +86,7 @@ config I2C_1
 	def_bool y
 
 if I2C_1
+
 config I2C_1_IRQ_PRI
```
Comment (ydamigos) [inline arch/x86/soc/intel_quark/quark_se/Kconfig.defconfig.series]: I am sure that I am missing something here. There is no interrupt property in intel,qmsi-i2c but there is IRQ_PRI here. 
Same for I2C_0.

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series [hunk 14]
```diff
@@ -99,6 +99,7 @@ config I2C_QMSI
 if I2C_QMSI
 config I2C_0
 	def_bool y
+
 config I2C_0_IRQ_PRI
```
Comment (ydamigos) [inline arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: I am sure that I am missing something here. There is no interrupt property in intel,qmsi-i2c but there is IRQ_PRI here.

File: dts/arc/quark_se_c1000_ss.dtsi [hunk 15]
```diff
@@ -61,5 +61,45 @@
 
 			status = "disabled";
 		};
+
+		i2c0: i2c@0x80012000 {
+			compatible = "intel,qmsi-ss-i2c";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0x80012000 0x400>;
```
Comment (ydamigos) [inline dts/arc/quark_se_c1000_ss.dtsi]: Is 0x400 correct here?

File: arch/arc/soc/quark_se_c1000_ss/Kconfig.defconfig [hunk 16]
```diff
@@ -162,9 +162,6 @@ if I2C_QMSI
 config I2C_0
 	def_bool y
 
-config I2C_0_NAME
-	default "I2C_2"
-
 config I2C_0_IRQ_PRI
```
Comment (ydamigos) [inline arch/arc/soc/quark_se_c1000_ss/Kconfig.defconfig]: I am sure that I am missing something here. There is no interrupt property in intel,qmsi-ss-i2c but there is IRQ_PRI here.
Same for I2C_1.

**Review by ydamigos — CHANGES_REQUESTED** [review]

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Indeed, sounds like some old left-over.

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Actuall, we both overlooked: it is there. Difference being that irq connection is done in init function.

Comment (ydamigos) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: I was talking about intel,qmsi-i2c.dtsi file. Will you add support to the i2c_qmsi driver to get IRQ configuration from DT.

File: arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series [hunk 17]
```diff
@@ -24,7 +24,7 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series]: Why not `if I2C_STM32`?

File: dts/x86/quark_x1000.dtsi (hunk 11 above)
Comment (ydamigos) [inline dts/x86/quark_x1000.dtsi]: Is IRQ_TYPE_LEVEL_LOW the IRQ priority?

File: arch/x86/soc/intel_quark/quark_x1000/Kconfig.defconfig.series (hunk 12 above)
Comment (ydamigos) [inline arch/x86/soc/intel_quark/quark_x1000/Kconfig.defconfig.series]: You can get IRQ_PRI from DT.

**Review by ydamigos — CHANGES_REQUESTED** [review]
Some of my comments were hidden as outdated so I added them again.

File: arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series (hunk 17 above)
Comment (tbursztyka) [reply arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series]: why? I did like it's done for SPI abve it.

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: aah, ok. I tried that but interrupts can't take more than 2 arguments there. Adding a interrupt-priority attribute could work, but not nice.

File: dts/x86/quark_x1000.dtsi (hunk 11 above)
Comment (tbursztyka) [reply dts/x86/quark_x1000.dtsi]: No that's the interrupt flags. I don't know if it's specific to ioapic.

File: arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series (hunk 17 above)
Comment (ydamigos) [reply arch/arm/soc/st_stm32/stm32f0/Kconfig.defconfig.series]: Sorry, I posted it in the wrong file.

File: arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series [hunk 18]
```diff
@@ -21,7 +21,7 @@ endif # GPIO_STM32
 
 if I2C && (I2C_1 || I2C_2 || I2C_3)
```
Comment (ydamigos) [inline arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series]: Why not if I2C_STM32? The same for the other series.

**Comment by ydamigos** [comment]
I could see a issue here. The second cell in interrupts property is used as IRQ priority in ARM device and as IRQ level in x86 (like in linux).

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (ydamigos) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Did you try to change #interrupt-cells number in intc node?

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Ah, that could help, thanks. I am new to dts, thus the trial & error path...

File: dts/arc/quark_se_c1000_ss.dtsi (hunk 15 above)
Comment (tbursztyka) [reply dts/arc/quark_se_c1000_ss.dtsi]: good catch, seems like it's 0x100

File: arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series (hunk 18 above)
Comment (tbursztyka) [reply arch/arm/soc/st_stm32/stm32f1/Kconfig.defconfig.series]: Ah ok, I see.

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Does not work, because of the interrupt-parent, &intc sets the interrupt-cells to 2.

Comment (ydamigos) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: I meant to change &intc node the interrupt-cells to 3 and update all the nodes using it (uart, i2c spi, ...).

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: oh, true. I need a break.

Will see that.

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Does not work, seems like a bug in extract_dts_includes.py to me but I am not entirely sure. (this tool _really_ needs to be rewritten)

**Comment by tbursztyka** [comment]
@galak, @ydamigos  could you re-review?

About the interrupt priority, it seems like an extract_dts_includes issue to me (among many others on that tool), so I won't address that in this PR.

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (ydamigos) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: I think you should modify also intel,ioapic.yaml from:
```
cell_string: IRQ

"#cells":
  - irq
  - sense
```
to:
```
cell_string: IRQ

"#cells":
  - irq
  - priority
  - sense
```
but I am not sure.

Comment (galak) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Does the intel,ioapic support priorities?  From what I can tell it doesn't seem to.  The binding for ioapic yaml is based on the linux binding if I remember correctly.

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: It does support priorities. I don't know why Linux does not expose this setting, could it be because the intel platforms supported there provide all the necessary information through PCI?

**Comment by tbursztyka** [comment]
@ydamigos , @galak I will make another PR for fixing this interrupt priority in intel's irq controllers, there are a lot of issues unrelated to i2c (uart not using CONFIG_ options for irq stuff, etc... it's a bit messy so it will require some cleanup patch first).

File: dts/bindings/i2c/snps,designware-i2c.yaml [hunk 19]
```diff
@@ -0,0 +1,36 @@
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+---
+title: DesignWare SPI node description
```
Comment (tkln) [inline dts/bindings/i2c/snps,designware-i2c.yaml]: Shouldn't this be I2C instead of SPI?

Comment (tbursztyka) [reply dts/bindings/i2c/snps,designware-i2c.yaml]: indeed, copy-pasted this from my other branch and forgot to change it

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml [hunk 20]
```diff
@@ -27,12 +27,6 @@ properties:
       generation: define
       category: required
 
-    interrupts:
```
Comment (ydamigos) [inline dts/bindings/i2c/st,stm32-i2c-v1.yaml]: Why did you remove this? ...

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml [hunk 21]
```diff
@@ -27,6 +27,12 @@ properties:
       generation: define
       category: required
 
+    interrupts:
```
Comment (ydamigos) [inline dts/bindings/i2c/st,stm32-i2c-v1.yaml]: And then you add it again.

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml (hunk 20 above)
Comment (tbursztyka) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml]: good question...

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (galak) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: Looking at the driver in zephyr, it doesn't seem that we allow or have means to set the priority in the ioapic (but I could be missing something).

Comment (tbursztyka) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: I am not really aware of the details in that part, but it seems that ioapic itself does not manage priorities directly, hardware wise. It seems to be handled in software through the ISR table.

**Comment by tbursztyka** [comment]
just a pushed a new version: I was using node names such as i2c0@0x.... instead of i2c0@....

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (andrewboie) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: We *do* set priorities. But it is not done in the IOAPIC. Priorities are reflected by the vector assignment in the IDT. Based on the priority level given to IRQ_CONNECT(), gen_idt uses this to find an appropriate range in the vector table to configure the interrupt. So the DTS configuration will need to express priorities.

**Comment by tbursztyka** [comment]
pushed again as master had an issue with ccs811

**Review by erwango — COMMENTED** [review]
Looks good to me a part from last comment from @ydamigos (interrupt property removal in stm32 i2c v1 yaml binding)

File: arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series (hunk 14 above)
Comment (galak) [reply arch/x86/soc/intel_quark/quark_d2000/Kconfig.defconfig.series]: > We do set priorities. But it is not done in the IOAPIC. Priorities are reflected by the vector assignment in the IDT. Based on the priority level given to IRQ_CONNECT(), gen_idt uses this to find an appropriate range in the vector table to configure the interrupt. So the DTS configuration will need to express priorities.

Trying to understand this, based on include/arch/x86/arch.h - it looks like MVIC doesn't support priority, and ioapic is done by how vector values are assigned.  Is that correct?

File: drivers/i2c/Kconfig.stm32 [hunk 22]
```diff
@@ -0,0 +1,48 @@
+#
+# Copyright (c) 2018 Intel Corporation
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+
+menuconfig I2C_STM32
```
Comment (ydamigos) [inline drivers/i2c/Kconfig.stm32]: STM32L0 series was added recently and it will be good to cleanup its Kconfig (arch/arm/soc/st_stm32/stm32l0/Kconfig.defconfig.series)

File: dts/bindings/i2c/st,stm32-i2c-v1.yaml (hunk 20 above)
Comment (ydamigos) [reply dts/bindings/i2c/st,stm32-i2c-v1.yaml]: We shouldn't remove this.

**Review by ydamigos — CHANGES_REQUESTED** [review]
I would propose to add clock-frequency = <I2C_BITRATE_FAST>; for all boards in the PR because most users will use 400 kHz bitrate. If you don't, just keep in mind that every user should create a board overlay to override it in application level.

**Comment by tbursztyka** [comment]
@ydamigos applied your comment. However, for the clock-frequency I applied what was existed previously i.e. the soc's dts (dts/<arch>/<soc>.dtsi is the on setting up the clock-frequency, and that seems to be always the standard speed. I don't see why I should change that, and if it has to: this change does not belong to this PR then. (it's not exactly the same subject).

**Comment by ydamigos** [comment]
> However, for the clock-frequency I applied what was existed previously i.e. the soc's dts (dts//.dtsi is the on setting up the clock-frequency, and that seems to be always the standard speed. I don't see why I should change that, and if it has to: this change does not belong to this PR then. (it's not exactly the same subject).

I was talking about board's dts files. You add i2c nodes to them in this PR so you could add a line to override the default.

**Comment by tbursztyka** [comment]
@ydamigos oh, yes, missed that.

**Review by ydamigos — APPROVED** [review]

**Comment by galak** [comment]
Merged PR #6610 which was the same as this with minor cleanups.
